internal/project/projectcollectionbuilder.go

```diff
@@ -0,0 +1,854 @@
+	// really even have an error condition until it tries to ask us language questions about
+	// a non-TS-handleable file.
+
+	if logger != nil {
+		elapsed := time.Since(startTime)
+		logger.Log(fmt.Sprintf("Completed file request for %s in %v", fileName, elapsed))
+	}
+}
+
+func (b *projectCollectionBuilder) DidUpdateATAState(ataChanges map[tspath.Path]*ATAStateChange, logger *logging.LogTree) {
+	updateProject := func(project dirty.Value[*Project], ataChange *ATAStateChange) {
+		project.ChangeIf(
+			func(p *Project) bool {
+				if p == nil {
+					return false
+				}
+				// Consistency check: the ATA demands (project options, unresolved imports) of this project
+				// has not changed since the time the ATA request was dispatched; the change can still be
+				// applied to this project in its current state.
+				return ataChange.TypingsInfo.Equals(p.ComputeTypingsInfo())
+			},
+			func(p *Project) {
+				// We checked before triggering this change (in Session.triggerATAForUpdatedProjects) that
+				// the set of typings files is actually different.
+				p.installedTypingsInfo = ataChange.TypingsInfo
+				p.typingsFiles = ataChange.TypingsFiles
+				fileWatchGlobs, directoryWatchGlobs := getTypingsLocationsGlobs(
+					ataChange.TypingsFilesToWatch,
+					b.sessionOptions.TypingsLocation,
+					p.currentDirectory,
+					b.fs.fs.UseCaseSensitiveFileNames(),
+				)
+				p.typingsFilesWatch = p.typingsFilesWatch.Clone(fileWatchGlobs)
+				p.typingsDirectoryWatch = p.typingsDirectoryWatch.Clone(directoryWatchGlobs)
+				p.dirty = true
+				p.dirtyFilePath = ""
+			},
+		)
+	}
+
+	for projectPath, ataChange := range ataChanges {
+		logger.Embed(ataChange.Logs)
+		if projectPath == inferredProjectName {
+			updateProject(b.inferredProject, ataChange)
+		} else if project, ok := b.configuredProjects.Load(projectPath); ok {
+			updateProject(project, ataChange)
+		}
+
+		if logger != nil {
+			logger.Log(fmt.Sprintf("Updated ATA state for project %s", projectPath))
+		}
+	}
+}
+
+func (b *projectCollectionBuilder) markProjectsAffectedByConfigChanges(
+	configChangeResult changeFileResult,
+	logger *logging.LogTree,
+) bool {
+	for projectPath := range configChangeResult.affectedProjects {
+		project, ok := b.configuredProjects.Load(projectPath)
+		if !ok {
+			panic(fmt.Sprintf("project %s affected by config change not found", projectPath))
+		}
+		project.ChangeIf(
+			func(p *Project) bool { return !p.dirty || p.dirtyFilePath != "" },
+			func(p *Project) {
+				p.dirty = true
+				p.dirtyFilePath = ""
+				if logger != nil {
+					logger.Logf("Marking project %s as dirty due to change affecting config", projectPath)
+				}
+			},
+		)
+	}
+
+	// Recompute default projects for open files that now have different config file presence.
+	var hasChanges bool
+	for path := range configChangeResult.affectedFiles {
+		fileName := b.fs.overlays[path].FileName()
+		_ = b.ensureConfiguredProjectAndAncestorsForOpenFile(fileName, path, logger)
```

Comment:
Computing this here seems like work that should be done when the next request comes in ? "you could get multiple watch events and its wasteful and expensive to do this right ?